extensions/amp-story/1.0/amp-story-embedded-component.js

```diff
@@ -536,37 +591,149 @@ export class AmpStoryEmbeddedComponent {
+    this.triggeringTarget_.classList.toggle(
+        'i-amphtml-expanded-component', false);
+    const embedId =
+      this.triggeringTarget_.getAttribute(EMBED_ID_ATTRIBUTE_NAME);
+
+    const embedStyleEl = dev().assertElement(embedStyleEls[embedId],
+        `Failed to look up embed style element with ID ${embedId}`);
+
+    const dataForExitExpandedMode = Object.assign({},
+        embedStyleEl[AMP_EMBED_DATA],
+        {transform: `scale(${embedStyleEl[AMP_EMBED_DATA].scaleFactor})`}
+    );
+    updateEmbedStyleEl(embedStyleEl, dataForExitExpandedMode);
+  }
+
+  /**
+   * Animates into expanded view. It calculates what the full-screen dimensions
+   * of the element will be, and uses them to deduce the translateX/Y values
+   * once the element reaches its full-screen size.
    * @param {!Element} target
    * @private
    */
   animateExpanded_(target) {
+    const embedId = target.getAttribute(EMBED_ID_ATTRIBUTE_NAME);
     const state = {};
+    const embedStyleEl = dev().assertElement(embedStyleEls[embedId],
+        `Failed to look up embed style element with ID ${embedId}`);
+    const embedData = embedStyleEl[AMP_EMBED_DATA];
     this.resources_.measureMutateElement(target,
         /** measure */
         () => {
           const targetRect = target./*OK*/getBoundingClientRect();
+          // TODO(#20832): Store DOMRect for the page in the store to avoid
+          // having to call getBoundingClientRect().
           const pageRect = this.componentPage_./*OK*/getBoundingClientRect();
 
-          const centeredTop = pageRect.height / 2 - targetRect.height / 2;
-          const centeredLeft = pageRect.width / 2 - targetRect.width / 2;
+          // Gap on the left of the element between full-screen size and
+          // current size.
+          const leftGap = (embedData.width - targetRect.width) / 2;
+          // Distance from left of page to what will be the left of the
+          // element in full-screen.
+          const fullScreenLeft = targetRect.left - leftGap - pageRect.left;
+          const centeredLeft = pageRect.width / 2 - embedData.width / 2;
+          state.translateX = centeredLeft - fullScreenLeft;
+
+          // Gap on the top of the element between full-screen size and
+          // current size.
+          const topGap = (embedData.height - targetRect.height) / 2;
+          // Distance from top of page to what will be the top of the element in
+          // full-screen.
+          const fullScreenTop = targetRect.top - topGap - pageRect.top;
+          const centeredTop = pageRect.height / 2 - embedData.height / 2;
+          state.translateY = centeredTop - fullScreenTop;
+        },
+        /** mutate */
+        () => {
+          target.classList.toggle('i-amphtml-expanded-component', true);
+
+          const dataForExpandedMode = Object.assign({}, embedData, {
+            transform: `translate3d(${state.translateX}px,
+              ${state.translateY}px, 0) scale(1)`,
+          });
```

Comment:
Any reason why you don't want to mutate `embedData`? Shouldn't it be always in sync with the styles?

Reply:
When we go to expanded view we set `scale(1)`, so we needed to keep that `scale` value intact in the `embedData` to know what scale to go back to when we exit the expanded view.

But I created a new `scaleFactor` property where we store this, so there shouldn't be a need to keep them out of sync anymore.